marimo/_server/api/endpoints/ws.py

```diff
@@ -0,0 +1,315 @@
+            # This can happen in local development when the client
+            # goes to sleep and wakes later. Just replace the session's
+            # socket, but keep its kernel
+            LOGGER.debug("Reconnecting session %s", session_id)
+            self._reconnect_session(session)
+        # Create a new session
+        else:
+            # If the client refreshed their page, there will be one
+            # existing session with a closed socket for a different session
+            # id; that's why we call `close_all_sessions`.
+            if mgr.mode == SessionMode.EDIT:
+                mgr.close_all_sessions()
+
+            mgr.create_session(
+                session_id=session_id,
+                session_handler=self,
+            )
+            self.status = ConnectionState.OPEN
+            # Let the frontend know it can instantiate the app.
+            await self._write_kernel_ready()
+
+        async def listen_for_messages() -> None:
+            while True:
+                (op, data) = await self.message_queue.get()
+                await self.websocket.send_text(
+                    json.dumps(
+                        {
+                            "op": op,
+                            "data": data,
+                        },
+                        cls=WebComponentEncoder,
+                    )
+                )
+
+        async def listen_for_disconnect() -> None:
+            try:
+                await self.websocket.receive_text()
+            except WebSocketDisconnect as e:
+                LOGGER.debug(
+                    "Websocket disconnected for session %s with exception %s",
+                    self.session_id,
+                    str(e),
+                )
+
+                # Change the status
+                self.status = ConnectionState.CLOSED
+
+                # When the websocket is closed, we wait TTL_SECONDS before
+                # closing the session. This is to prevent the session from
+                # being closed if the during an intermittent network issue.
+                if self.manager.mode == SessionMode.RUN:
+
+                    def _close() -> None:
+                        if self.status != ConnectionState.OPEN:
+                            LOGGER.debug(
+                                "Closing session %s (TTL EXPIRED)",
+                                self.session_id,
+                            )
+                            self.manager.close_session(self.session_id)
+
+                    session = self.manager.get_session(self.session_id)
+                    cancellation_handle = asyncio.get_event_loop().call_later(
+                        Session.TTL_SECONDS, _close
+                    )
+                    if session is not None:
+                        self.cancel_close_handle = cancellation_handle
+
+                # Stop listen_for_messages
+                listen_for_messages_task.cancel()
```

Comment:
@mscolnick 

in `RUN` mode, should this task be kept alive until the TTL expires? i.e., should the task be cancel in the `_close` function (line 237)